### ISSUE TITLE
Open styleable shadow tree and theming

### ISSUE THREAD

**Issue by rniwa** [w3c/tpac2024-breakouts#27]
### Session description

Session to discuss open styleable shadow tree and theming https://github.com/WICG/webcomponents/issues/909

### Session goal

Come up with a concrete proposal for open styleable shadow tree and theming

### Who can attend

Anyone may attend (Default)

### IRC channel (Optional)

#webcomponents

### Other sessions where we should avoid scheduling conflicts (Optional)

#26

**Comment by sorvell** [comment]
This is an attempt to frame the discussion of this breakout session. Some of the ideas here should likely be broken out into separate issues, but the issues for [open styleable](https://github.com/WICG/webcomponents/issues/909) and [theming](https://github.com/WICG/webcomponents/issues/864) are already over-saturated.

### Background
Discussion on open styleable has drifted away from the original proposal focused on providing a bridge for legacy styling and towards a focus on theming. However, theming based on applying user provided stylesheets sacrifices Shadow DOM's styling encapsulation. A proper theming solution should preserve style encapsulation in general while allowing authors to expose a themeable CSS-based API. If this session can focus on developing a good theming system, perhaps the simpler initial proposal for open styleable will be more appealing.

### Goals
* focus on theming, not open-styleable: the former may unblock the later.
* explore building on `parts`
  * opt-in for making parts deeply selectable 
    * should support combinators, e.g. like the old shadow/deep selectors but limited to the "part tree"
    * must be declarative, e.g. new mode or flag or via CSS `appearance: custom;`
  * fixing [exportparts](https://github.com/WICG/webcomponents/issues/1051)
  * allowing a host to participate without defining a part or using its tag name
* ensure new/all CSS features adhere to https://github.com/w3c/csswg-drafts/issues/1995
  * helps avoid problems like inconsistent behavior of [:has](https://drafts.csswg.org/selectors-4/#relational)
  * makes sure new features like [mixins](https://github.com/w3c/csswg-drafts/issues/9350) can be used
### Non-goals
* discuss open-styleable, instead consider this blocked on the theming discussion
* discuss css property filtering since this is a general problem applicable to any pseudo-element selector

**Comment by dandclark** [comment]
Notes from the breakout session:
Wednesday, September 25th, 2024

1:52 PM 
— Zakim sees Michael_Warren, justinf, westbrook on the speaker queue
1:52 PM <dandclark> Dan Clark 
sorvell: Thanks!
1:52 PM <justinf> 
can someone repost the link?
1:52 PM <dandclark> Dan Clark 
masonf: The specific proposal is `part` exposes part of the tree but it painful to use, so this @@@. 
1:53 PM <masonf> 
ack Michael
1:53 PM 
— Zakim sees justinf, westbrook on the speaker queue
1:53 PM <TabAtkins> 
s/@@@/exposes all the parts as a reduced version of the shadow tree, selectable with combinators/
1:53 PM <dandclark> Dan Clark 
Michael_Warren0: It's tempting to think about the problem and solution from component perspective. But it's multidimensional. Depends on who you are relative to the cmponent and the application that's using it
1:53 PM <justinf> 
lea do you ahve the proposal link?
1:53 PM <lea> Lea Verou 
justinf: https://github.com/w3c/csswg-drafts/issues/10939
1:53 PM 
— astearns https://github.com/w3c/csswg-drafts/issues/10939 justinf 
1:53 PM <justinf> 
thanks!
1:54 PM <dandclark> Dan Clark 
... @@@ isn't necessarily theming because I'm not just changing a color, there are other bigger things like layout I'm changing.
1:54 PM 
...E.g. Shoelace, author has no knowledge  of what exists in the component
1:54 PM <rniwa> 
q?
1:54 PM 
— Zakim sees justinf, westbrook on the speaker queue
1:55 PM <masonf> 
ack justin
1:55 PM 
— Zakim sees westbrook on the speaker queue
1:55 PM <dandclark> Dan Clark 
...Want to narrow down the relationship between the component and the environment.
1:55 PM <westbrook> Westbrook Johnson 
q-
1:55 PM 
— Zakim sees no one on the speaker queue
1:55 PM <sorvell> Steve Orvell 
Thanks so much Lea, this proposal is amazing (https://github.com/w3c/csswg-drafts/issues/10939)
1:55 PM <dandclark> Dan Clark 
justinf: To rniwa's point about concrete proposals. We had thing that exist, CSS props and parts
1:55 PM 
...upcoming CSS features like mixins, vague versions of `theme`. Each has pros and cons.
1:55 PM 
...gatherting that might help us make a rubric for evaluating other things.
1:56 PM <sorvell> Steve Orvell 
q+
1:56 PM 
— Zakim sees sorvell on the speaker queue
1:56 PM <dandclark> Dan Clark 
...Important aspect of the solution is you can potentially merge different styles.
1:56 PM <masonf> 
ack sorvell
1:56 PM 
— Zakim sees no one on the speaker queue
1:57 PM <dandclark> Dan Clark 
...If you go through these you will come up with the types of objectsions you'll need to just solutions by
1:57 PM 
sorvell: We use style isolation in shadowDom as big hammer for core problem with CSS 
1:57 PM 
...Want page to have styling policy where I say some user of lib can style certain things. But can't style others.
1:57 PM 
E.g. input has placeholder. Lots of props you can set on placeholder that break the input.
1:58 PM 
s/E.g./...e.g.
1:58 PM 
...If browser had way to limit what you can do with that, that would be good
1:58 PM <lea> Lea Verou 
thanks sorvell, glad I posted it then! Took me a while to press the submit button, I thought it was so outside the Overton window and so ambitious that I'd lose all credibility 🤣
1:58 PM <dandclark> Dan Clark 
...It's just a CSS concept we don't have.
1:58 PM <masonf> 
q?
1:58 PM 
— Zakim sees no one on the speaker queue
1:58 PM <dandclark> Dan Clark 
...For lack of better word I'm calling it a policy
1:58 PM <justinf> 
q+
1:58 PM 
— Zakim sees justinf on the speaker queue
1:58 PM <masonf> 
ack justin
1:58 PM 
— Zakim sees no one on the speaker queue
1:59 PM <dandclark> Dan Clark 
justinf: Looking at nonwebcomonent solutions as well. It's common in corporate environments it's common to have policy that bans certain selectors. Have to use opaque classnames.
1:59 PM 
...That's common with solutions like userland CSS modules
1:59 PM 
...Points to people not using selectors like they could.
2:00 PM 
...Maybe also indicates we don't need the full selector feature set.
2:00 PM 
...People get by now using IDs and opaque classnames.
2:00 PM <masonf> 
q?
2:00 PM 
— Zakim sees no one on the speaker queue
2:00 PM <gregwhitworth> 
q+
2:00 PM 
— Zakim sees gregwhitworth on the speaker queue
2:00 PM <dandclark> Dan Clark 
...I worry about exposing tree structure relationship of parts. Hvae seen components change where that changes, so that's a breaking change for the component.
2:00 PM <gregwhitworth> 
ack gregwhitworth
2:00 PM 
— Zakim sees no one on the speaker queue
2:00 PM <sorvell> Steve Orvell 
[+q](https://www.irccloud.com/irc/w3.org/channel/%2Bq)
2:00 PM 
— Zakim sees sorvell on the speaker queue
2:00 PM <masonf> 
ack sorvell
2:00 PM 
— Zakim sees no one on the speaker queue
2:01 PM <masonf> 
q+
2:01 PM 
— Zakim sees masonf on the speaker queue
2:01 PM <JRJurman> 
q+
2:01 PM 
— Zakim sees masonf, JRJurman on the speaker queue
2:01 PM <rniwa> 
q+
2:01 PM 
— Zakim sees masonf, JRJurman, rniwa on the speaker queue
2:01 PM <dandclark> Dan Clark 
sorvell: I think this is a big space. Have webcomponents CG but don't have CSS horsepower in that group. Want more participation. Know there's interest in CSSWG but not sure how we can do that. Ideas for that would be awesome
2:01 PM <sanketj> Sanket Joshi 
q+
2:01 PM 
— Zakim sees masonf, JRJurman, rniwa, sanketj on the speaker queue
2:01 PM <justinf> 
q+
2:01 PM 
— Zakim sees masonf, JRJurman, rniwa, sanketj, justinf on the speaker queue
2:01 PM <dandclark> Dan Clark 
masonf: ONe specific idea was `::part` with some combinator. Can we look at that more?
2:02 PM <masonf> 
ack mason
2:02 PM 
— Zakim sees JRJurman, rniwa, sanketj, justinf on the speaker queue
2:02 PM <masonf> 
ack jrjurman
2:02 PM 
— Zakim sees rniwa, sanketj, justinf on the speaker queue
2:02 PM <dandclark> Dan Clark 
JRJurman: Thinking about what greg mentioned, I get the sense that as people put out components, if there was some way to expose some part for styling, there will probably be pressure to expose the whole thing. E.g. extensions that want to change everything. 
2:03 PM 
...Not sure if I'd ever want to say I"m making a thing and only some can be styled, then anticipate github issues about wanting more to be styleable
2:03 PM 
...If people want to style it, there's pressure to let them
2:03 PM <masonf> 
ack rniwa
2:03 PM 
— Zakim sees sanketj, justinf on the speaker queue
2:03 PM <dandclark> Dan Clark 
rniwa: focusing on theming use cases might be valuable way to limit discussion. Problem space is v big.
2:04 PM 
...Despite I said needs to be solution for all of them, maybe don't need solution for all use cases.
2:04 PM <sorvell> Steve Orvell 
q+
2:04 PM 
— Zakim sees sanketj, justinf, sorvell on the speaker queue
2:04 PM <masonf> 
ack sanket
2:04 PM 
— Zakim sees justinf, sorvell on the speaker queue
2:04 PM <dandclark> Dan Clark 
masonf: +1 let;s pick solution and see what cases it solves
2:04 PM 
sanketj: I agree. Let's outline what proposals we have, what they solve.
2:04 PM <justinf> 
I'd like to present an idea for ::theme() that I think addresses a lot of problems with parts
2:04 PM <Michael_Warren0> 
+1 to Lea's proposal above. It seems like it would work for headless components (ie, in a tailwind global css env) and would work for design systems like web awesome
2:05 PM <dandclark> Dan Clark 
...Maybe there's some lines we can continue to bring discussion forward
2:05 PM <masonf> 
ack justin
2:05 PM 
— Zakim sees sorvell on the speaker queue
2:05 PM <justinf> 
https://docs.google.com/presentation/d/13GJAJ_1WwxwTmusXnfobveDdYdaOr1TW-YRPhCzLDtg/edit#slide=id.g2f8ba040b01_0_0
2:06 PM <dandclark> Dan Clark 
justinf: Tab and I looked at theme. Big problem was it selected too many things. People that added part attrs to their shadow maybe dind't want those to be exposed all the way up the tree
2:06 PM 
...Component using child component doesn't necessarily want to expose all the child parts for theming
2:06 PM 
...So wanted to make theming more targeteable.
2:07 PM <Michael_Warren0> 
to justin's point, a good thing about lea's proposal above is that it doesnt need names that would have conflicts
2:07 PM <dandclark> Dan Clark 
...Also need a way to filter by host, but not by host tag name, might be different by component. Want host to opt into exposing things as theme, and establish some identifier for those. When selecting for those, select some theme name and part.
2:07 PM <Michael_Warren0> 
the point about having naming conflicts is a good one, so imo a solution should prevent that
2:07 PM <dandclark> Dan Clark 
...Goes deeply down the page, is coherent with parts but more targeted.
2:08 PM 
...Goes a long way towards letting design systems do what they want
2:08 PM <masonf> 
ack sorvell
2:08 PM 
— Zakim sees no one on the speaker queue
2:08 PM <dandclark> Dan Clark 
sorvell: I think that's interesting idea. 
2:08 PM <TabAtkins> 
I'd love to see that written as a rough proposal, it would help me understand it a little better.
2:08 PM <dandclark> Dan Clark 
...The reason I like having solution that includes combinators is my mental model for sytling is needs to satisfy: want to style one, want to style all, want to style some, where typically that's some subtree
2:08 PM <lea> Lea Verou 
+1 to sorvell
2:09 PM <dandclark> Dan Clark 
...Main point is: we have exportparts. We have parts. Not great, cumbersome -- don't work because need to export parts on all the things, becomes burdensome. 
2:09 PM <TabAtkins> 
+1 to this
2:09 PM <dandclark> Dan Clark 
...Adding wildcards is very simple way to achieve something useful
2:09 PM 
justinf: It's more broken because of name collisions
2:09 PM <westbrook> Westbrook Johnson 
https://github.com/WICG/webcomponents/issues/1051#issuecomment-2375196189
2:10 PM <dandclark> Dan Clark 
TabAtkins: Forwarding and prefix-changing was part of original proposal
2:10 PM 
sorvell: It's completely unusable. Shoelace uses but that's about it because it's so cumbersome
2:10 PM <masonf> 
q?
2:10 PM 
— Zakim sees no one on the speaker queue
2:10 PM <TabAtkins> 
exportparts="*" to forward all as-is, exportparts="foo-*: bar-*" to forward all the foo- prefixes, replaced with bar- prefixes
2:10 PM <dandclark> Dan Clark 
masonf: Seems like separate issue, but a good one
2:11 PM <kschmi> 
https://github.com/w3c/csswg-drafts/issues/10939
2:12 PM <masonf> 
Go here https://www.w3.org/events/meetings/19813be0-9902-4bad-8d50-3ad49ba792e6/
2:12 PM <Michael_Warren0> 
export parts seems tricky. if lea's proposal gets traction, i could see export parts being deprecated in favor of that feature. lea's proposal seems to be a re-imagining of parts in a more robust way
2:12 PM <masonf> 
and click the join button
2:13 PM <dandclark> Dan Clark 
sorvell: I beg your patience because I respec the people in this room. One of the things as a component user is when there's awesome new stuff added CSS, but we wonder if it will work in shadow DOM.
2:13 PM 
...The more we can make stuff so these questions are worked out the better. E.g. how does `has` work with shadow DOM is unclear
2:13 PM <Michael_Warren0> 
+1000 to awesome css features in shadow dom
2:13 PM <dandclark> Dan Clark 
...The more we can get a genral solution the better
2:14 PM 
lea: <presents her new proposal>
2:14 PM 
...Filed this yesterday, wasn't sure if should post
2:15 PM 
...Was thinking part is a huge pain point both when defining and using components. When defining because have to expose with reasonable names. When using you have to figure out names, build mental model of how fits together, and there's still stuff you can't do
2:15 PM 
...I was thinking what is the real author intent? It's that we want to expose subset of shadow tree with `part`.
2:15 PM 
...When people use part, over 90% of elements are exposed
2:15 PM 
...Usually need reason *not* to expose something
2:16 PM 
...Open styleable sort of solves this but usually still need to hide something
2:16 PM 
...What if we had html attribute -- tentatively `export`; it just exports that element
2:16 PM 
...Introduce combinator that gives you access to this exposed subtree. Don't know if you need mutiple or just once works
2:16 PM <justinf> 
q+
2:16 PM 
— Zakim sees justinf on the speaker queue
2:16 PM <dandclark> Dan Clark 
...That subtee only consists of exposed elements.
2:16 PM <sorvell> Steve Orvell 
q+
2:16 PM 
— Zakim sees justinf, sorvell on the speaker queue
2:17 PM 
⇐ giacomo-petri quit (~giacomopetri@9d40a032.publics.cloak)
2:17 PM <dandclark> Dan Clark 
...Within it, selectors just work normally but only apply to that subtree
2:17 PM 
...You could even use attribute selectors on `part` , making `part` work better
2:17 PM 
...Can introduce more granularity, cutting out subtrees, or even granularity on sub-element level
2:18 PM <Michael_Warren0> 
q+
2:18 PM 
— Zakim sees justinf, sorvell, Michael_Warren on the speaker queue
2:18 PM <emilio> 
q+
2:18 PM 
— Zakim sees justinf, sorvell, Michael_Warren, emilio on the speaker queue
2:18 PM <dandclark> Dan Clark 
...Can even introduce syntax to allow aria to hook in
2:18 PM 
...Or other parts of the platform. Depends on the overlap of what elements you want to expose for styling and for other purposes
2:18 PM 
...Can also be used with native elements, instead of introducing a bunch of new pseudos
2:19 PM 
...Maybe the element type would not be exposed so you can't use type selectors
2:19 PM 
...Could even end up with a different structure then what you originally had
2:19 PM 
...<walks though example with spinner, where exposed subteee is different>
2:19 PM 
...But it could break expectations e.g. if you're applying grid
2:20 PM 
...But part has the same problem
2:20 PM <bkardell_> Brian Kardell 
q+
2:20 PM 
— Zakim sees justinf, sorvell, Michael_Warren, emilio, bkardell_ on the speaker queue
2:20 PM <miriam> Miriam Suzanne 
q+
2:20 PM 
— Zakim sees justinf, sorvell, Michael_Warren, emilio, bkardell_, miriam on the speaker queue
2:20 PM <bkardell_> Brian Kardell 
q-
2:20 PM 
— Zakim sees justinf, sorvell, Michael_Warren, emilio, miriam on the speaker queue
2:20 PM <rniwa> 
q?
2:20 PM 
— Zakim sees justinf, sorvell, Michael_Warren, emilio, miriam on the speaker queue
2:20 PM <dandclark> Dan Clark 
...Open questions: Has access to full flattened subtree? And what if the better approach is to opt out rather than opt in?
2:21 PM <masonf> 
ack justin
2:21 PM 
— Zakim sees sorvell, Michael_Warren, emilio, miriam on the speaker queue
2:21 PM 
⇐ jarhar and kschmi quit  
2:21 PM <miriam> Miriam Suzanne 
q-
2:21 PM 
— Zakim sees sorvell, Michael_Warren, emilio on the speaker queue
2:21 PM <dandclark> Dan Clark 
justinf: think we need virtual DOM in this other place, where component can produce virtual subtree.
2:21 PM 
...One detail is that custom element tagnames aren't reliable as selectors. Host needs to apply a stable label.
2:22 PM 
...Material-button could be under any tagname. For this type of component, I want to expose this type of thing inside of it. That applies deeply
2:22 PM 
lea: You can just use part
2:22 PM 
justinf: But that's applied from the outside
2:22 PM <emilio> 
q-
2:22 PM 
— Zakim sees sorvell, Michael_Warren on the speaker queue
2:23 PM <masonf> 
ack sorvell
2:23 PM 
— Zakim sees Michael_Warren on the speaker queue
2:23 PM 
⇐ sabidussi_marco quit (~sabidussimarco@9d40a032.publics.cloak)
2:24 PM <dandclark> Dan Clark 
sorvell: I like the direction of this. Exploring stuff like how do you want to export this is good. Might have internal classes I don't want to expose, but fine to export others. Other thing i want to expore is it seems like we're circling this idea of a policy of how I want to expose styling. Does this idea need to be locked in shadow DOM?
2:24 PM 
...I'm describing what I want to be styleable. That's sort of what scope is doing
2:24 PM <masonf> 
ack michael
2:24 PM 
— Zakim sees no one on the speaker queue
2:24 PM <miriam> Miriam Suzanne 
Scope is selector-in. It can't handle this sort of thing
2:24 PM 
This sort of policy has to be DOM-out
2:25 PM <dandclark> Dan Clark 
Michael_Warren0: I think about this a lot -- stable names. Concern I have is applying that selector uses something considered internal to that component if it were a 3p. That introduces brittleness.
2:25 PM 
...The component could change, that thing we exposed might not be the same anymore. Could break every selector. Needs to be something that isn't selector based. Maybe proposal doesn
2:25 PM 
...'t let you use those. It's a place that can introduce brittleness. Class name might not exist later.
2:26 PM <sorvell> Steve Orvell 
Rough... `@scope ... { appearance: exports-only; ... }`
2:26 PM <dandclark> Dan Clark 
masonf: Adding this export to your shadow means you're adding it to contract of your component
2:26 PM <astearns> Alan Stearns 
q?
2:26 PM 
— Zakim sees no one on the speaker queue
2:26 PM <dandclark> Dan Clark 
justinf: Might have classes that are internal that you don't want to be exposed. @@@
2:27 PM 
Michael_Warren0: I might refactor the component in a way that I don't know if it's breaking, because I don't know who's querying it.
2:27 PM 
...Might want to thing about putting stable names back in this.
2:27 PM <miriam> Miriam Suzanne 
sorvell: Happy to follow up on that at some point (an issue?) but I'm not sure what you see that scoped appearance property doing.
2:27 PM <dandclark> Dan Clark 
...Want to make them truly stable so they can't be changed
2:27 PM 
⇐ masonf quit (~masonf@9d40a032.publics.cloak)
2:28 PM <dandclark> Dan Clark 
lea: I heard discussions about theming. that's what the next breakout is about, but on a lower level
2:28 PM <astearns> Alan Stearns 
https://www.w3.org/events/meetings/a9540089-cb3a-413b-966a-8c1034b31b11/ for the possible next session
